__tests__/ship.test.js

```diff
@@ -0,0 +1,293 @@
+                            _ships: [],
+                            addShip: jest.fn(),
+                            removeShip: jest.fn()
+                        },
+                        {
+                            name: 'St Petersburg',
+                            _ships: [],
+                            addShip: jest.fn(),
+                            removeShip: jest.fn()
+                        }
+                    ]
+                }
+            };
+            balticExplorer = new Ship(balticExplorerConfig);
+        });
+
+        afterEach(() => {
+            balticExplorer = null;
+        });
+
+        test('Throws an error if "_itinerary.ports" is empty', () => {
+            const error = 'The ship has no itinerary.';
+
+            balticExplorer._itinerary.ports = [];
+            expect(() => {
+                balticExplorer._setSail();
+            }).toThrowError(error);
+        });
+
+        test('Throws an error if "_currentPort" is null', () => {
+            const error = 'The ship has already set sail.';
+
+            balticExplorer._currentPort = null;
+            expect(() => {
+                balticExplorer._setSail();
+            }).toThrowError(error);
+        });
+
+        test('Throws an error if "_itinerary.ports" has one port AND "_previouslyDockedPorts" is not empty', () => {
+            const error = 'The ship has already set sail.';
+
+            balticExplorer._itinerary.ports = [jest.fn()];
+            balticExplorer._previouslyDockedPorts.push(jest.fn());
+            expect(() => {
+                balticExplorer._setSail();
+            }).toThrowError(error);
+        });
+
+        test('State of "_currentPort" should be set to null', () => {
+            expect(balticExplorer._currentPort).not.toBeNull();
+            balticExplorer._setSail();
+            expect(balticExplorer._currentPort).toBeNull();
+        });
+
+        test('Remove one port from "_itinerary.ports"', () => {
+            expect(balticExplorer._itinerary.ports.length).toBe(2);
+            expect(balticExplorer._itinerary.ports[0].name).toBe('Helsinki');
+            expect(balticExplorer._itinerary.ports[1].name).toBe('St Petersburg');
+            balticExplorer._setSail();
+            expect(balticExplorer._itinerary.ports.length).toBe(1);
+            expect(balticExplorer._itinerary.ports[0].name).toBe('St Petersburg');
+        });
+
+        test('removeShip should be called when _setSail is invoked', () => {
+            balticExplorer._setSail();
+            const [departedPort] = balticExplorer._previouslyDockedPorts;
+            expect(departedPort.removeShip).toHaveBeenCalledWith(balticExplorer);
+        });
+
+        test('Add the removed port from "_itinerary.ports" to "_previouslyDockedPorts"', () => {
+            expect(balticExplorer._previouslyDockedPorts.length).toBe(0);
+            expect(balticExplorer._previouslyDockedPorts).toEqual([]);
```

Comment:
Do these need to be here?

Reply:
Yeah I see your point on these, and the similar ones above, they are quite redundant